File: functions/deployment.js
const axios = require("axios");
const logger = require("firebase-functions/logger");

/**
 * Restore deleted doc.
 * @param {object} event Firestore event
 */
async function restoreTriggerDoc(event) {
  const verRef = event.data.ref;
  await verRef.set({createdAt: new Date()});
  logger.info(`restored doc ${verRef.path}`);
}

/**
 * Upgrade system on deployment.
 * @param {object} db Firestore
 * @param {object} auth FirebaseAuth
 */
async function upgradeData(db, auth) {
  const confRef = db.collection("service").doc("conf");
  const confDoc = await confRef.get();
  const currentVersion = confDoc.exists ? confDoc.data().dataVersion : 0;
  logger.info(`current data version: ${currentVersion}.`);

  const ts = new Date();
  const createdAt = ts;
  const updatedAt = ts;

  if (!confDoc.exists) {
    await confRef.set({
      dataVersion: 0,
      uiVersion: "",
      createdAt,
      updatedAt,
    });
  }

  const dataVersion = 1;

  if (currentVersion === dataVersion) {
    logger.info("Skip to upgrade data.");
    return;
  }

  if (currentVersion === 0) {
    logger.info("upgrade data 0 to 1.");

    const testRef = db.collection("orgs").doc("test");

    const user = await auth.createUser({
      email: process.env.PRIMARY_USER_EMAIL,
    });

    await db.collection("service").doc("admins").set({
      name: "Administrators",
      accounts: [user.uid],
      createdAt,
      updatedAt,
    });

    await testRef.set({
      name: "Test",
      accounts: [user.uid],
      createdAt,
      updatedAt,
    });

    const userDoc = await testRef.collection("users").add({
      name: "Primary user",
      email: process.env.PRIMARY_USER_EMAIL,
      createdAt,
      updatedAt,
    });

    await testRef.collection("accounts").doc(user.uid).set({
      user: userDoc.id,
      createdAt,
      updatedAt,
    });

    await testRef.collection("groups").doc("managers").set({
      name: "Managers",
      users: [userDoc.id],
      createdAt,
      updatedAt,
    });
  }

  logger.info(`set data version: ${dataVersion}.`);
  await confRef.update({
    dataVersion,
    createdAt,
    updatedAt,
  });
}

/**
 * Set UI version.
 * @param {object} event Firestore event
 * @param {object} db Firestore
 */
async function setUiVersion(event, db) {
  const ts = new Date();
  const url = `https://${event.project}.web.app/version.json`;
  const params = `check=${ts.getTime()}`;
  const res = await axios.get(`${url}?${params}`);
  const {version} = res.data;
  logger.info(`get new UI version: ${version}`);

  const confRef = db.collection("service").doc("conf");
  const confDoc = await confRef.get();

  if (confDoc.data().uiVersion === version) return;

  await confRef.update({
    uiVersion: version,
    updatedAt: ts,
  });
  logger.info(`updated ${confRef.path}.uiVersion: ${version}`);
}

module.exports = {
  restoreTriggerDoc,
  upgradeData,
  setUiVersion,
};

File: functions/deployment.py
from datetime import datetime
from firebase_admin import auth
from firebase_functions.firestore_fn import Event, DocumentSnapshot
from google.cloud import firestore
import os
import requests


def restore_trigger_doc(
    event: Event[DocumentSnapshot],
):
    print("Start: restoreTriggerDoc")
    ref = event.data.reference
    ref.set(
        {
            "createdAt": firestore.SERVER_TIMESTAMP,
        }
    )
    print(f"restored doc: {ref.path}")
    print("End  : restoreTriggerDoc")


def set_ui_version(
    event: Event[DocumentSnapshot],
):
    print("Start: setUiVersion")
    res = requests.get(
        f"https://{event.project}.web.app/version.json"
        f"?check={datetime.now().timestamp()}"
    )
    if res.status_code == 200:
        ver = res.json()["version"]
        doc = event.data
        if doc.get("uiVersion") != ver:
            doc.reference.update(
                {
                    "uiVersion": ver,
                    "updatedAt": firestore.SERVER_TIMESTAMP,
                }
            )
    else:
        print(f"Error: HTTP Status: {res.status_code}")
    print("End  : setUiVersion")


def upgrade_data(
    db: firestore.Client,
    auth: auth.Client,
):
    print("Start: upgrade_data")

    conf_ref = db.collection("service").document("conf")
    conf_doc = conf_ref.get()
    cur_ver = 0

    if conf_doc.exists:
        cur_ver = conf_doc.get("dataVersion")
    else:
        conf_ref.set(
            {
                "dataVersion": 0,
                "uiVersion": "",
                "policy": "## Privacy policy",
                "createdAt": firestore.SERVER_TIMESTAMP,
                "updatedAt": firestore.SERVER_TIMESTAMP,
            }
        )
    print(f"CurVer: {cur_ver}")

    new_ver = 1
    print(f"NewVer: {new_ver}")

    if cur_ver == new_ver:
        pass
    elif cur_ver == 0:
        user = auth.create_user(
            email=os.environ.get("PRIMARY_USER_EMAIL"),
        )
        db.collection("service").document("admins").set(
            {
                "name": "Administrators",
                "accounts": [user.uid],
                "createdAt": firestore.SERVER_TIMESTAMP,
                "updatedAt": firestore.SERVER_TIMESTAMP,
            }
        )

        test_ref = db.collection("orgs").document("test")

        test_ref.set(
            {
                "name": "Test",
                "accounts": [user.uid],
                "createdAt": firestore.SERVER_TIMESTAMP,
                "updatedAt": firestore.SERVER_TIMESTAMP,
            }
        )

        (_, user_doc) = test_ref.collection("users").add(
            {
                "name": "Primary user",
                "email": os.environ.get("PRIMARY_USER_EMAIL"),
                "createdAt": firestore.SERVER_TIMESTAMP,
                "updatedAt": firestore.SERVER_TIMESTAMP,
            }
        )

        test_ref.collection("accounts").document(user.uid).set(
            {
                "user": user_doc.id,
                "createdAt": firestore.SERVER_TIMESTAMP,
                "updatedAt": firestore.SERVER_TIMESTAMP,
            }
        )

        test_ref.collection("groups").document("managers").set(
            {
                "name": "Managers",
                "users": [user_doc.id],
                "createdAt": firestore.SERVER_TIMESTAMP,
                "updatedAt": firestore.SERVER_TIMESTAMP,
            }
        )

        conf_ref.update(
            {
                "dataVersion": new_ver,
                "createdAt": firestore.SERVER_TIMESTAMP,
                "updatedAt": firestore.SERVER_TIMESTAMP,
            }
        )

    print("End  : upgrade_data")
